File: src/test/java/com/example/pages/OrderSuccessPage.java
package com.example.pages;

import com.example.utilities.ConfigReader;
import io.qameta.allure.Step;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;

/**
 * Page Object for order success page (payment_done).
 * Contains: Download Invoice, Continue (data-qa="continue-button").
 */
public class OrderSuccessPage extends BasePage {

    private static final By ORDER_SUCCESS_MESSAGE =
            By.xpath("//*[contains(.,'Your order has been placed successfully!')]");

    // <a href="/download_invoice/500" class="btn btn-default check_out">Download Invoice</a>
    private final By downloadInvoiceLink =
            By.cssSelector("a.btn.btn-default.check_out[href^='/download_invoice']");
    // <a href="/" data-qa="continue-button" class="btn btn-primary">Continue</a>
    private final By continueButton =
            By.cssSelector("a.btn.btn-primary[data-qa='continue-button']");

    public OrderSuccessPage(WebDriver driver) {
        super(driver);
    }

    @Step("Click Download Invoice")
    public void clickDownloadInvoice() {
        scrollIntoView(downloadInvoiceLink);
        click(downloadInvoiceLink);
    }

    @Step("Click Continue on order success page")
    public void clickContinue() {
        clickViaJavaScript(continueButton);
    }

    /** Checks if Continue button is visible and clickable. */
    public boolean isContinueButtonVisible() {
        if (!isElementPresent(continueButton)) {
            return false;
        }
        WebElement el = getElement(continueButton);
        return el.isDisplayed() && el.isEnabled();
    }

    /** Waits until Continue button is clickable. */
    public boolean waitUntilContinueClickable() {
        try {
            wait.until(ExpectedConditions.elementToBeClickable(continueButton));
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    /** Checks if Download Invoice link is visible. */
    public boolean isDownloadInvoiceVisible() {
        return isElementPresent(downloadInvoiceLink);
    }

    /**
     * Waits for order success (redirect to payment_done or success message on page).
     * Uses orderSuccessWaitTimeout from config.
     */
    public boolean waitForOrderSuccessMessageOrPaymentDone() {
        int timeout = Integer.parseInt(ConfigReader.get("orderSuccessWaitTimeout", "15"));
        WebDriverWait w = new WebDriverWait(driver, Duration.ofSeconds(timeout));
        try {
            w.until(d -> {
                if (d.getCurrentUrl() != null && d.getCurrentUrl().contains("payment_done")) {
                    return true;
                }
                return !d.findElements(ORDER_SUCCESS_MESSAGE).isEmpty();
            });
            return true;
        } catch (Exception e) {
            return driver.getCurrentUrl() != null && driver.getCurrentUrl().contains("payment_done");
        }
    }
}

File: pages/order_success_page.py
"""Order success page (payment_done) – Download Invoice, Continue."""
from selenium.webdriver.common.by import By
from selenium.webdriver.remote.webdriver import WebDriver
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import WebDriverWait

from pages.base_page import BasePage
from utilities.config_reader import get


ORDER_SUCCESS_MESSAGE = (
    By.XPATH,
    "//*[contains(.,'Your order has been placed successfully!')]"
)


class OrderSuccessPage(BasePage):
    download_invoice_link = (
        By.CSS_SELECTOR,
        "a.btn.btn-default.check_out[href^='/download_invoice']"
    )
    continue_button = (By.CSS_SELECTOR, "a.btn.btn-primary[data-qa='continue-button']")

    def __init__(self, driver: WebDriver):
        super().__init__(driver)

    def click_download_invoice(self) -> None:
        self.scroll_into_view(self.download_invoice_link)
        self.click(self.download_invoice_link)

    def click_continue(self) -> None:
        self.click_via_javascript(self.continue_button)

    def is_continue_button_visible(self) -> bool:
        if not self.is_element_present(self.continue_button):
            return False
        el = self.get_element(self.continue_button)
        return el.is_displayed() and el.is_enabled()

    def wait_until_continue_clickable(self) -> bool:
        try:
            self.wait.until(EC.element_to_be_clickable(self.continue_button))
            return True
        except Exception:
            return False

    def is_download_invoice_visible(self) -> bool:
        return self.is_element_present(self.download_invoice_link)

    def wait_for_order_success_message_or_payment_done(self) -> bool:
        timeout = int(get("orderSuccessWaitTimeout", "15"))
        w = WebDriverWait(self.driver, timeout)
        try:
            def _check(d):
                if d.current_url and "payment_done" in d.current_url:
                    return True
                return len(d.find_elements(*ORDER_SUCCESS_MESSAGE)) > 0
            w.until(_check)
            return True
        except Exception:
            return bool(self.driver.current_url and "payment_done" in self.driver.current_url)
